README.md
# praat-textgrids -- Praat TextGrid manipulation in Python

## Description

`textgrids` is a module for handling Praat TextGrid files in any format (short text, long text, or binary). The module implements five classes, from largest to smallest:

* `TextGrid` -- a `dict` with tier names as keys and `Tier`s as values
* `Tier` -- a `list` of either `Interval` or `Point` objects
* `Interval` -- an `object` representing Praat intervals
* `Point` -- a `namedtuple` representing Praat points
* `Transcript` -- a `str` with special methods for transcription handling

All Praat text objects are represented as `Transcript` objects.

The module also exports the following variables:

* `diacritics` -- a `dict` of all diacritics with their Unicode counterparts
* `inline_diacritics` -- a `dict` of inline (symbol-like) diacritics
* `index_diacritics` -- a `dict` of over/understrike diacritics
* `symbols` -- a `dict` of special Praat symbols with their Unicode counterparts
* `vowels` -- a `list` of all vowels in either Praat or Unicode notation

And the following constants (although they CAN be changed due to Python they SHOULDN’T be changed):

* `BINARY` -- symbolic name for the binary file format
* `TEXT_LONG` -- symbolic name for the long text file format
* `TEXT_SHORT` -- symbolic name for the short text file format
* `version` -- module version as string

## Version

This file documents `praat-textgrids` version 1.3.1.

## Copyright

Copyright © 2019–20 Legisign.org, Tommi Nieminen <software@legisign.org>

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see <https://www.gnu.org/licenses/>.

## Module contents

### 0. Module properties

Besides `textgrids.version`, which contains the module version number as string, the module exports the following properties:

#### 0.1. symbols

`symbols` is a `dict` that contains all the Praat special notation symbols (as keys) and their Unicode counterparts (as values).

#### 0.2. vowels

`vowels` is a `list` of all vowel symbols in either Praat notation (e.g., `"\as"`) or in Unicode. It is used by `Interval` methods `containsvowel()` and `startswithvowel()`, so changing it, for example, adding new symbols to it or removing symbols used for other purposes in a specific case, will change how those methods function.

#### 0.3. diacritics, inline_diacritics, and index_diacritics

`diacritics` is a `dict` of all diacritics in Praat notation (as keys) and their Unicode counterparts (as values).

`inline_diacritics` and `index_diacritics` are subsets of `diacritics`. The former are semantically diacritics but appear as inline symbols, the latter are the “true” diacritics (i.e., under- or overstrikes) that need special handling when transcoding.

### 0.4. TEXT_LONG, TEXT_SHORT, BINARY

Symbolic constants specifying different file formats in `TextGrid.format()` and `TextGrid.write()` methods. Internally they are just small integers (0, 1, and 2, respectively). The default format is `TEXT_LONG`.

### 1. TextGrid

`TextGrid` is an `collections.OrderedDict` whose keys are tier names (strings) and values are `Tier` objects. The constructor takes an optional filename argument for easy loading and parsing textgrid files.

#### 1.1. Properties

All the properties of `dict` plus:

* `filename` holds the textgrid filename, if any. `read()` and `write()` methods both set or update it.

#### 1.2. Methods

All the methods of `dict` plus:

* `parse()` -- parse string into a TextGrid
* `read()` -- read (and parse) a TextGrid file
* `tier_from_csv()` -- read a textgrid tier from a CSV file
* `tier_to_csv()` -- write a textgrid tier into a CSV file
* `write()` -- write a TextGrid file

`parse()` takes an obligatory string (or `bytes`) argument which contains textgrid data in any of Praat’s three formats (long text, short text, or binary).

`read()` and `write()` both take an obligatory filename argument.

`write()` can take an optional argument specifying the file format; this can be one of `BINARY` (= `int` 2), `TEXT_LONG` (= `int` 0, the default), or `TEXT_SHORT` (= `int` 1).

`tier_from_csv()` and `tier_to_csv()` both take two obligatory arguments, the tier name and the filename, in that order.

### 2. Tier

`Tier` is a list of either `Interval` or `Point` objects.

**NOTE:** `Tier` only allows adding `Interval` or `Point` objects. Adding anything else or mixing `Interval`s and `Point`s will trigger an exception.

#### 2.2. Properties

All the properties of `list` plus:

* `is_point_tier` -- `bool` `True` for point tier, `False` for interval tier.
* `tier_type` -- `str`, either `"IntervalTier"` or `"PointTier"`

`tier_type` exists principally for the convenience of the formatting functions.

#### 2.3. Methods

All the methods of `list` plus:

* `concat()` -- concatenate intervals
* `to_csv()` -- convert tier data into a CSV-like list

`concat()` concatenates given intervals into one. It takes two arguments, `first=` and `last=`, both of which are integer indexes with the usual Python semantics: 0 stands for the first element, -1 for the last element, these being also the defaults. The function raises a `TypeError` if used with a point tier, and `ValueError` if the parameters do not specify a valid slice. **Note** that this is a function and returns the result instead of modifying the Tier in place.

`to_csv()` returns a CSV-like list. It’s mainly intended to be used from the `TextGrid` level method `tier_to_csv()` but can be called directly if writing to a file is not desired.

### 3. Interval

`Interval` is an `object` class representing one Interval on an IntervalTier.

#### 3.1. Properties

* `dur` -- interval duration (`float`)
* `mid` -- interval midpoint (`float`)
* `text` -- text label (`Transcript`)
* `xmax` -- interval end time (`float`)
* `xmin` -- interval start time (`float`)

#### 3.3. Methods

* `containsvowel()` -- does the interval contain a vowel?
* `endswithvowel()` -- does the interval end with a vowel?
* `startswithvowel()` -- does the interval start with a vowel?
* `timegrid()` -- create a time grid

`containsvowel()`, `endswithvowel()`, and `startswithvowel()` are `bool` functions. They check for possible vowels in the `text` property in both Praat notation and Unicode, but can of course make an error if symbols are used in an unexpected way. They don’t take arguments. (Internally, `endswithvowel()` first transcodes the text to IPA removing all diacritics to simplify the test.)

`timegrid()` returns a list of timepoints (in `float`) evenly distributed from `xmin` to `xmax`. It takes an optional integer argument specifying the number of timepoints desired; the default is 3. It raises a `ValueError` if the argument is not an integer or is less than 1.

### 4. Point

`Point` is a `namedtuple` representing one Point on a PointTier.

#### 4.1. Properties

* `text` -- text label (`Transcript`)
* `xpos` -- temporal position (`float`)

### 5. Transcript

`Transcript` is a `str`-derived class with one special method: `transcode()`.

### 5.1. Properties

All the properties of `str`.

#### 5.2. Methods

All the methods of `str` plus:

* `transcode()` -- convert Praat notation to Unicode or vice versa.

Without arguments, `transcode()` assumes its input to be in Praat notation and converts it to Unicode; no check is made as to whether the input really is in Praat notation but nothing **should** happen if it isn’t. User should take care and handle any exceptions.

Optional `to_unicode=False` argument inverts the direction of the transcoding from Unicode to Praat. Again, it is not checked whether input is in Unicode.

With optional `retain_diacritics=True` argument the transcoding does not remove over- and understrike diacritics from the result.

## Examples

### Snippet 1: list syllable durations

    import sys
    import textgrids

    for arg in sys.argv[1:]:
        # Try to open the file as textgrid
        try:
            grid = textgrids.TextGrid(arg)
        # Discard and try the next one
        except:
            continue

        # Assume "syllables" is the name of the tier
        # containing syllable information
        for syll in grid['syllables']:
            # Convert Praat to Unicode in the label
            label = syll.text.transcode()
            # Print label and syllable duration, CSV-like
            print('"{}";{}'.format(label, syll.dur))

### Snippet 2: convert any textgrid to binary format

    import sys
    import os.path
    import textgrids

    for arg in sys.argv[1:]:
        name, ext = os.path.splitext(arg)
        try:
            grid = textgrids.TextGrid(arg)
        except (textgrids.ParseError, textgrids.BinaryError):
            print('Not a recognized file format!', file=sys.stderr)
            continue

        # Write a new file
        grid.write(name + '.bin', fmt=textgrids.BINARY)


HISTORY.txt
  2019-06-27  1.0     r12 declared as 1.0.
  2019-06-30  1.0.1   Bug fix. Vowels lacked 'aeiouyæø'!
  2019-06-30  1.0.2   Further bug fix. Added in-line diacritics to symbols.
  2019-07-01  1.0.3   More verbose docstrings. Removed optional argument from
                      TextGrid.tier_from_csv().
  2019-07-10  1.1.0   When retain_diacritics=True, Transcript.transcode() now
                      works as it should: in Praat notation, over/understrikes
                      follow the symbol, in Unicode, they precede it. This
                      requires some swapping of characters. This also means
                      that besides ‘diacritics’ variable there are also
                      ‘inline_diacritics’ and ‘index_diacritics’.
  2019-07-11  1.1.1   Separated Transcript in a submodule.
  2019-07-14  1.2.0   Read binary textgrids.
  2020-01-21  1.3.0   Added TextTier as a alternative name for PointTier.
                      AFAIK TextTiers cannot be created in Praat but they do
                      appear in Praat’s documentation so I guess we should
                      take care of them?
  2020-03-23  1.3.1   BUG FIX: Tier.concat() fixed.


pitches.py
#!/usr/bin/env python3
'''

  pitches.py -- an extremely simple Praat Pitch object parser.

  2019-01-07    Started. (TN)

'''

from collections import namedtuple

Measurement = namedtuple('Measurement', ['t', 'f0'])

class Pitch(list):
    def __init__(self, filename=None):
        self.filename = filename
        if self.filename:
            self.read(filename)

    def __str__(self):
        return '\n'.join(['{:.3f};{:.2f}'.format(t, f0) for t, f0 in self])

    def read(self, filename):
        def getval(s):
            return float(s.split(' = ')[1])

        self.filename = filename
        with open(filename, 'r') as infile:
            i = 0
            skip = True
            for line in infile:
                if line.startswith('xmin ='):
                    xmin = getval(line)
                elif line.startswith('xmax ='):
                    xmax = getval(line)
                elif line.startswith('dx = '):
                    dx = getval(line)
                elif 'candidate [1]' in line:
                    skip = False
                elif not skip and 'frequency =' in line:
                    freq = getval(line)
                    skip = True
                    self.append(Measurement(xmin + i * dx, freq))
                    i += 1

if __name__ == '__main__':
    import sys

    for arg in sys.argv[1:]:
        pitch = Pitch(arg)
        print(pitch)


textgrids/templates.py
'''

  templates.py

  2019-07-15    Done.
  2019-08-02    Bug fix. Always TEXT_{LONG|SHORT}, never
                {LONG|SHORT}_TEXT.
  2019-08-04    Bug fix (line breaks, indentation in templates).

'''

# Symbolic names for the formats

TEXT_LONG = 0
TEXT_SHORT = 1
BINARY = 2

# Formatting templates

long_header = '''File type = "ooTextFile"
Object class = "TextGrid"

xmin = {}
xmax = {}
tiers? <exists>
size = {}
item []:'''

short_header = '''File type = "ooTextFile"
Object class = "TextGrid"

{xmin}
{xmax}
<exists>
{length}
'''

long_tier = '''
    item [{}]:
        class = "{}"
        name = "{}"
        xmin = {}
        xmax = {}
        {}: size = {}'''

short_tier = '''"{tier_type}"
"{name}"
{xmin}
{xmax}
{length}
'''

long_point = '''
            points [{}]:
                xpos = {}
                text = "{}"'''

short_point = '''{xpos}
"{text}"
'''

long_interval = '''
            intervals [{}]:
                xmin = {}
                xmax = {}
                text = "{}"'''

short_interval = '''{xmin}
{xmax}
"{text}"
'''


textgrids/__init__.py
#!/usr/bin/env python3
'''textgrids -- Read and handle Praat’s textgrid files

  © Legisign.org, Tommi Nieminen <software@legisign.org>, 2012-20
  Published under GNU General Public License version 3 or newer.

  2020-03-27  1.4.0.dev1    New ideas from Maxine Fily: tier + tier should
                            work as expected, concatenating tiers.
  2020-03-29  1.4.0.dev2    Fixed Tier.to_csv() bug, but the changes in dev1
                            still need testing.
  2020-03-29  1.4.0.dev3    For tier + tier to work, tier.xmin and tier.xmax
                            should be read after all. Parser changes as well
                            as fixes for Tier.__add__().

'''

import codecs
import io
import struct
from collections import OrderedDict, namedtuple
from .transcript import *
from .templates import *

# Global constant

version = '1.4.0.dev3'

class BinaryError(Exception):
    '''Read error for binary files.'''
    pass

class ParseError(Exception):
    '''Read error for text files.'''
    def __str__(self):
        return 'Parse error on line {}'.format(self.args[0])

# Point class
Point = namedtuple('Point', ['text', 'xpos'])

class Interval(object):
    '''Interval is a timeframe xmin..xmax labelled "text".

    xmin, xmax are floats, text a string that is converted into a Transcript
    object.
    '''

    def __init__(self, text=None, xmin=0.0, xmax=0.0):
        self.text = Transcript(text)
        self.xmin = xmin
        self.xmax = xmax
        if self.xmin > self.xmax:
            return ValueError

    def __repr__(self):
        '''Return (semi-)readable representation of self.'''
        return '<Interval text="{}" xmin={} xmax={}>'.format(self.text,
                                                             self.xmin,
                                                             self.xmax)

    def containsvowel(self):
        '''Boolean: Does the label contain a vowel?'''
        global vowels
        return any([vow in self.text for vow in vowels])

    @property
    def dur(self):
        '''Return duration of the Interval.'''
        return self.xmax - self.xmin

    def endswithvowel(self):
        '''Boolean: does the label end with a vowel?

        Discards diacritics before testing.'''
        global vowels
        text = self.text.transcode(retain_diacritics=False)
        return any([text.endswith(vow) for vow in vowels])

    @property
    def mid(self):
        '''Return temporal midpoint for the Interval.'''
        return self.xmin + self.dur / 2

    def startswithvowel(self):
        '''Boolean: does the label start with a vowel?'''
        global vowels
        return any([self.text.startswith(vow) for vow in vowels])

    def timegrid(self, num=3):
        '''Create an even-spaced time grid.

        Takes one optional integer argument "num", the number of timepoints.
        Returns a list of timepoints (floats) where the first element is
        Interval. xmin and the last Interval.xmax.
        '''
        if num <= 1 or num != int(num):
            raise ValueError('value not integer or <= 1')
        step = self.dur / num
        return [self.xmin + (step * i) for i in range(num + 1)]

class Tier(list):
    '''Tier is a list of either Interval or Point objects.'''

    def __init__(self, data=None, xmin=0.0, xmax=0.0, point_tier=False):
        self.is_point_tier = point_tier
        if not data:
            self.xmin = xmin
            self.xmax = xmax
            data = []
        else:
            self.xmin = data[0].xmin
            self.xmax = data[-1].xmax
        super().__init__(data)

    def __add__(self, tier):
        '''Concatenate tiers.'''
        # Only a tier or the empty list can be concatenated with a tier
        if tier and not isinstance(tier, Tier):
            raise TypeError('incompatible types')
        # Concatenate only tiers of the same type
        if self.is_point_tier != tier.is_point_tier:
            raise TypeError('tier types differ')
        # Sanity check
        if self.xmax > tier.xmin:
            raise ValueError('time values do not match')
        return Tier(super().__add__(tier))

    def merge(self, first=0, last=-1):
        '''Merge intervals Tier[first]...Tier[last].

        Parameters follow the usual Python index semantics: starting from 0,
        negative number count from the end. Note that the tier is not changed
        in place, you have to assign it to a variable.

        The method raises a TypeError if the Tier is a point tier,
        and ValueError if Tier[first:last] is not a slice.
        '''
        if self.is_point_tier:
            raise TypeError
        area = self[first:last + 1]
        if not area:
            raise ValueError
        xmin = self[first].xmin
        xmax = self[last].xmax
        text = ''.join([segm.text for segm in area])
        new = Interval(text, xmin, xmax)
        self = self[:first] + [new] + self[last + 1:]
        return self

    def to_csv(self):
        '''Format tier data as CSV, each row a separate string.'''
        if self.is_point_tier:
            return ['"{}";{}'.format(p.text, p.xpos) for p in self]
        else:
            return ['"{}";{};{}'.format(i.text.transcode(),
                                        i.xmin,
                                        i.xmax) for i in self]

    @property
    def tier_type(self):
        '''Return tier type as string (for convenience).'''
        return 'PointTier' if self.is_point_tier else 'IntervalTier'

class TextGrid(OrderedDict):
    '''TextGrid is a dict of tier names (keys) and Tiers (values).'''

    def __init__(self, filename=None):
        self.xmin = self.xmax = 0.0
        self.filename = filename
        if self.filename:
            self.read(self.filename)

    def __repr__(self):
        '''Return Praat (long) text format representation'''
        return self.format()

    def format(self, fmt=TEXT_LONG):
        '''Format data as long text, short text, or binary.

        The optional argument can be TEXT_LONG (or 0) for long text,
        TEXT_SHORT (or 1) for short text, or BINARY (or 2) for binary.
        '''
        global BINARY, TEXT_LONG, TEXT_SHORT
        if fmt == TEXT_LONG:
            return(self._format_long())
        elif fmt == TEXT_SHORT:
            return(self._format_short())
        elif fmt == BINARY:
            return(self._format_binary())
        else:
            raise ValueError

    def _format_binary(self):
        '''Format self as binary. Not intended to be used directly.'''
        out = b'ooBinaryFile\x08TextGrid'
        out += struct.pack('>2d?i', self.xmin, self.xmax, True, len(self))
        for name, tier in self.items():
            typ = b'PointTier' if tier.is_point_tier else b'IntervalTier'
            out += struct.pack('B', len(typ))
            out += typ
            try:
                encoded_name = bytes(name, encoding='ascii')
                out += struct.pack('>h', len(encoded_name))
            except UnicodeEncodeError:
                encoded_name = bytes(name, encoding='utf-16-be')
                out += struct.pack('>2h', -1, len(encoded_name))
            out += encoded_name
            out += struct.pack('>2di', self.xmin, self.xmax, len(tier))
            for elem in tier:
                if tier.is_point_tier:
                    out += struct.pack('>d', elem.xpos)
                else:
                    out += struct.pack('>2d', elem.xmin, elem.xmax)
                try:
                    encoded_text = bytes(elem.text, encoding='ascii')
                    out += struct.pack('>h', len(encoded_text))
                except UnicodeEncodeError:
                    encoded_text = bytes(elem.text, encoding='utf-16-be')
                    out += struct.pack('>2h', -1, len(encoded_text))
                out += encoded_text
        return out

    def _format_long(self):
        '''Format self as long text. Not intended to be used directly.'''
        global long_header, long_tier, long_point, long_interval
        out = long_header.format(self.xmin, self.xmax, len(self))
        tier_count = 1
        for name, tier in self.items():
            if tier.is_point_tier:
                tier_type = 'PointTier'
                elem_type = 'points'
            else:
                tier_type = 'IntervalTier'
                elem_type = 'intervals'
            out += long_tier.format(tier_count,
                                    tier_type,
                                    name,
                                    self.xmin,
                                    self.xmax,
                                    elem_type,
                                    len(tier))
            for elem_count, elem in enumerate(tier, 1):
                if tier.is_point_tier:
                    out += long_point.format(elem_count,
                                             elem.xpos,
                                             elem.text)
                else:
                    out += long_interval.format(elem_count,
                                                elem.xmin,
                                                elem.xmax,
                                                elem.text)
        return out

    def _format_short(self):
        '''Format self as short text. Not intended to be used directly.'''
        global short_header, short_tier, short_point, short_interval
        out = short_header.format(xmin=self.xmin,
                                  xmax=self.xmax,
                                  length=len(self))
        for name, tier in self.items():
            out += short_tier.format(tier_type=tier.tier_type,
                                     name=name,
                                     xmin=self.xmin,
                                     xmax=self.xmax,
                                     length=len(tier))
            for elem in tier:
                if tier.is_point_tier:
                    out += short_point.format(xpos=elem.xpos,
                                              text=elem.text)
                else:
                    out += short_interval.format(xmin=elem.xmin,
                                                 xmax=elem.xmax,
                                                 text=elem.text)
        return out

    def parse(self, data):
        '''Parse textgrid data.

        Obligatory argument "data" is bytes.
        '''
        if not isinstance(data, bytes):
            raise TypeError
        binary = b'ooBinaryFile\x08TextGrid'
        text = ['File type = "ooTextFile"', 'Object class = "TextGrid"', '']
        # Check and then discard binary header
        if data[:len(binary)] == binary:
            buff = io.BytesIO(data[len(binary):])
            try:
                self._parse_binary(buff)
            except (IndexError, ValueError):
                raise BinaryError
        else:
            coding = 'utf-8'
            # Note and then discard BOM
            if data[:2] == b'\xfe\xff':
                coding = 'utf-16-be'
                data = data[2:]
            # Now convert to a text buffer
            buff = [s.strip() for s in data.decode(coding).split('\n')]
            # Check and then discard header
            if buff[:len(text)] != text:
                raise TypeError
            buff = buff[len(text):]
            # If the next line starts with a number, this is a short textgrid
            if buff[0][0] in '-0123456789':
                self._parse_short(buff)
            else:
                self._parse_long(buff)

    def _parse_binary(self, data):
        '''Parse BINARY textgrid files. Not intended to be used directly.'''
        sBool, sByte, sShort, sInt, sDouble = [struct.calcsize(c) for c in '?Bhid']

        self.xmin, self.xmax = struct.unpack('>2d', data.read(2 * sDouble))
        if not struct.unpack('?', data.read(sBool))[0]:
            return

        tiers = struct.unpack('>i', data.read(sInt))[0]
        for i in range(tiers):
            size = struct.unpack('B', data.read(sByte))[0]
            desc = data.read(size)
            # TextTiers don’t appear anywhere in Praat’s UI as far as
            # I can see but they are mentioned in Praat’s online docs.
            # TextTier is just a PointTier with a different name
            # so let’s treat it as one.
            if desc in (b'PointTier', b'TextTier'):
                point_tier = True
            elif desc == b'IntervalTier':
                point_tier = False
            else:
                raise BinaryError
            tier = Tier(point_tier=point_tier)
            size = struct.unpack('>h', data.read(sShort))[0]
            tier_name = data.read(size).decode()
            xmin = data.read(sDouble)
            xmax = data.read(sDouble)
            elems = struct.unpack('>i', data.read(sInt))[0]
            for j in range(elems):
                if point_tier:
                    xpos = struct.unpack('>d', data.read(sDouble))[0]
                else:
                    xmin, xmax = struct.unpack('>2d', data.read(2 * sDouble))
                size = struct.unpack('>h', data.read(sShort))[0]
                # Apparently size -1 is an index that UTF-16 follows
                if size == -1:
                    size = struct.unpack('>h', data.read(sShort))[0] * 2
                    coding = 'utf-16-be'
                else:
                    coding = 'ascii'
                text = Transcript(data.read(size).decode(coding))
                if point_tier:
                    tier.append(Point(text, xpos))
                else:
                    tier.append(Interval(text, xmin, xmax))
            self[tier_name] = tier
            self.xmin = xmin
            self.xmax = xmax

    def _parse_long(self, data):
        '''Parse LONG textgrid files. Not intended to be used directly.'''
        grab = lambda s: s.split(' = ')[1]
        self.xmin, self.xmax = [float(grab(s)) for s in data[:2]]
        if data[2] != 'tiers? <exists>':
            return
        tiers = int(grab(data[3]))
        p = 6
        for i in range(tiers):
            tier_type, tier_name = [grab(s).strip('"') for s in data[p:p + 2]]
            tier = Tier(point_tier=(tier_type != 'IntervalTier'))
            p += 2
            tier.xmin, tier.xmax = [float(grab(s)) for s in data[p:p + 2]]
            tier_len = int(grab(data[p + 2]))
            p += 4
            for j in range(tier_len):
                if tier.is_point_tier:
                    x1 = float(grab(data[p]))
                    text = Transcript(grab(data[p + 1]).strip('"'))
                    tier.append(Point(text, x1))
                    p += 3
                else:
                    x0, x1 = [float(grab(s)) for s in data[p:p + 2]]
                    text = Transcript(grab(data[p + 2]).strip('"'))
                    tier.append(Interval(text, x0, x1))
                    p += 4
            self[tier_name] = tier

    def _parse_short(self, data):
        '''Parse SHORT textgrid files. Not intended to be used directly.'''
        self.xmin, self.xmax = [float(s) for s in data[:2]]
        if data[2] != '<exists>':
            return
        tiers = int(data[3])
        p = 4
        for i in range(tiers):
            tier_type, tier_name = [s.strip('"') for s in data[p:p + 2]]
            tier = Tier(point_tier=(tier_type == 'PointTier'))
            tier.xmin, tier.xmax = [float(x) for x in data[p + 2:p + 4]]
            elems = int(data[p + 4])
            p += 5
            for j in range(elems):
                if tier.is_point_tier:
                    x1, text = data[p:p + 2]
                    x1 = float(x1)
                    text = Transcript(text.strip('"'))
                    tier.append(Point(text, x1))
                    p += 2
                else:
                    x0, x1, text = data[p:p + 3]
                    x0 = float(x0)
                    x1 = float(x1)
                    text = Transcript(text.strip('"'))
                    tier.append(Interval(text, x0, x1))
                    p += 3
            self[tier_name] = tier

    def read(self, filename):
        '''Read given file as a TextGrid.

        "filename" is the name of the file.
        '''
        self.filename = filename
        with open(self.filename, 'rb') as infile:
            data = infile.read()
        self.parse(data)

    def tier_from_csv(self, tier_name, filename):
        '''Import CSV file to an interval or point tier.

        "tier_name" (string) is the name for the new tier.
        "filename" (string) is the name of the input file.
        '''
        import csv
        tier = None
        with open(filename, 'r') as csvfile:
            reader = csv.reader(csvfile, delimiter=';')
            for line in reader:
                if len(line) == 2:
                    text, xpos = line
                    elem = Point(text, xpos)
                elif len(line) == 3:
                    text, xmin, xmax = line
                    elem = Interval(text, xmin, xmax)
                else:
                    raise ValueError
                if not tier:
                    if isinstance(elem, Point):
                        tier = Tier(point_tier=True)
                    else:
                        tier = Tier()
                # The following may raise a TypeError
                tier.append(elem)
        self[tier_name] = tier

    def tier_to_csv(self, tier_name, filename):
        '''Export given tier into a CSV file.

        "tier_name" is the name of the (existing) tier to be exported.
        "filename" is the name of the file.
        '''
        with open(filename, 'w') as csvfile:
            csvfile.write('\n'.join(self[tier_name].to_csv()))

    def write(self, filename, fmt=TEXT_LONG):
        '''Write the text grid into a Praat TextGrid file.

        Obligatory argument "filename" gives the file name.
        Optional argument "formatting" can be TEXT_LONG, TEXT_SHORT, or
        BINARY (see "TextGrid.format()"). Default is TEXT_LONG.
        '''
        global BINARY, TEXT_SHORT, TEXT_LONG
        with open(filename, 'w' if fmt != BINARY else 'wb') as outfile:
            outfile.write(self.format(fmt))
